app/src/main/java/ryey/easer/skills/usource/airplane_mode/AirplaneModeTracker.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2016 - 2022 Rui Zhao <renyuneyun@gmail.com>
+ *
+ * This file is part of Easer.
+ *
+ * Easer is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Easer is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Easer.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package ryey.easer.skills.usource.airplane_mode;
+
+import android.app.PendingIntent;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.provider.Settings;
+
+import androidx.annotation.NonNull;
+
+import ryey.easer.skills.condition.SkeletonTracker;
+
+public class AirplaneModeTracker extends SkeletonTracker<AirplaneModeUSourceData> {
+
+    private final AirplaneModeUSourceData condition;
+    private boolean curMode;
+
+    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            if (Intent.ACTION_AIRPLANE_MODE_CHANGED.equals(intent.getAction())) {
+                updateTrackerMode(intent.getBooleanExtra("state", false));
+            }
+        }
+    };
+    private static final IntentFilter intentFilter = new IntentFilter(Intent.ACTION_AIRPLANE_MODE_CHANGED);
+
+    AirplaneModeTracker(Context context, AirplaneModeUSourceData data,
+                        @NonNull PendingIntent event_positive,
+                        @NonNull PendingIntent event_negative) {
+        super(context, data, event_positive, event_negative);
+        this.condition = data;
+    }
+
+    @Override
+    public void start() {
+        context.registerReceiver(broadcastReceiver, intentFilter);
+        curMode = (Settings.System.getInt(context.getContentResolver(), Settings.System.AIRPLANE_MODE_ON, 0) != 0);
+        newSatisfiedState(condition.match(curMode));
```

Comment:
The checking and setting of initial state had better be done in the constructor.

Reply:
Is there no situation where `stop()` gets called after initialization, followed by another `start()` some time later? I assumed that this may happen if the condition is disabled and re-enabled afterwards. In that case we avoid issues when changing the airplane mode setting while the condition is disabled.
This is all based on the assumption that this can happen. I did not research this possibility thoroughly though.

Reply:
Further testing revealed that my assumption was incorrect. This still leaves the question why we need a `start()` method, but that's not relevant for this particular PR.

Reply:
I believe your thoughts are valid, but it does not happen for the implementation. In the current implementation (which should be improved!), the Tracker is instantiated once when the service starts, and starts listening for changes. In fact, the `stop()` is not really called unless the service restarts or reloads. (See [ConditionHolderService.java](https://github.com/renyuneyun/Easer/blob/master/app/src/main/java/ryey/easer/core/ConditionHolderService.java))

Previously there was a method to check initial states. But some conditions do not have a sensible way to retrieve initial states, so that method was removed afterwards, and the checking is put to inside the constructor where needed. As `satisfied` is `Boolean` which can be `null`, it is still alright if a condition can't check state initially.


Some more rationale: The mechanism for trackers was designed in this way because I initially wanted to make the tracker more resource efficient -- to have one tracker for one type of condition, and use it to listen to all changes of the conditions of the same type. That's why there comes the `ConditionHolderService`, rather than making them inside a `Lotus` in `EHService`. But due to time issues, I ended up with something more similar to the slots for events. Was planning to change it, but never find enough big block of time to do that.